Add release notes for 0.273 (#17775)
prestodb/presto: presto-docs/src/main/sphinx/release/release-0.273.rst

=============
Release 0.273
=============

**Highlights**
==============

**Details**
===========

General Changes
_______________

> * Fix Disaggregated Coordinator resource group aggregation issue related to WAITING_FOR_PREREQUISITE queries.

Reviewer: - lower case 'coordinator'?
- can we explain what is the exact issue exposed to the users?

Author: done and added some details about the issue.

> * Fix Iceberg $files table in case of column dropping.

Reviewer 2: probably needs to move into Iceberg connector section

Author: moved to iceburg section

* Fix a bug where ``EXPLAIN (TYPE IO)`` did not include the index source from an index join.
* Fix a memory overcounting issue for using varchar types that could cause queries to fail with exceeded memory limit errors.

> * Fix an error in ``tdigest_agg``.

Reviewer 2: Functions have to be written with  `:func:`

Reviewer: Can we elaborate more about the "error"?

Author: add some description for the error:
Fix an off-by-one error during buffer size calculation in :func:`tdigest_agg`.

> * Add `to_base32` and `from_base32` varbinary functions.

Reviewer 2: Functions have to be written with `:func:`
See [this](https://github.com/prestodb/presto/commit/e4dd2048b38e2677792415bb561e639cb05bd9eb#diff-31566b652020a06db29e535cd1943a102fb3a44b568963879ff1e05130a5fb96R7) as an example.

Author: done

> * Add support for TRUNCATE TABLE statement.

Reviewer: Not sure if TRUNCATE TABLE needs quotes

Author: updated

> * Add support for TRUNCATE TABLE statement.

Reviewer 3: Truncate table should be in double back-tick quotes like

```
``TRUNCATE TABLE``
```

Should also link to relevant documentation here.

Author: added quotes but I can't find the document for "truncate" in https://prestodb.io/docs/current/
is that something that has to be added ? @rschlussel

Author: I checked the commit and it does include a doc change adding TRUNCATE. But we can't add a link here before the docs are published ?

* Add support to set timeout on the query analyzer runtime to prevent long running query analysis like complex regular expressions. The timeout duration can be set by the configuration property `planner.query-analyzer-timeout` or session property `query_analyzer_timeout`.

> * Add trim_array function.

Reviewer: Functions have to be written with :func:. Also explain what this function does.

Author: done:
* Add :func:`trim_array` function that can delete elements from the end of an ordinary array.

> * Add trim_array function.

Reviewer 3: should use :func:

Author: updated

> * Added a new optimization for showing results for (interactive) distinct limit N as they become available with no buffering.  This can be enabled using the session param: quick_distinct_limit_enabled to true.

Reviewer: do we need quotes for quick_distinct_limit_enabled?

Author: updated

> * Removed state (PageBuilder) in ZIP_WITH as it was getting clobbered when used in REDUCE_AGG context.

Reviewer: What is the user-facing problem here? More efficient using the function?

> * Removed state (PageBuilder) in ZIP_WITH as it was getting clobbered when used in REDUCE_AGG context.

Reviewer 3: ```
Improve performance of :func: `zip_with` when used inside a :func:`reduce_agg`.
```

Author: Updated

* ...
* ...

> * ...

Reviewer 2: stray lines

Author: will remove them

Author: removed

> * Bump postgresql driver to 42.3.3.

Reviewer 4: Upgrade postgresql driver to 42.3.3.

Author: corrected

* Change  protoc and grpc version in presto-grpc-api module to adapt linux aarch64.

> * If hive files are changed, the modification time can be used to invalidate parquet metadata cache immediately.

Reviewer 3: should go in the Hive section.  And "hive" should be capitalize (Hive).

Author: updated. and moved it to the hive section

> * Upgrade zstandard compression to version 1.5.2.2. This can improved about 2% overall CPU with data compressed with zstandard.

Reviewer 3: This improves the cpu used for data compressed with zstandard by about 2%.

Author: updated.

> * We have an option to now to do a local round robin shuffle to reduce the partial distinct limit output size.

Reviewer: Are these general changes or connector-specific changes? They look to be misc items...

> * We have an option to now to do a local round robin shuffle to reduce the partial distinct limit output size.

Reviewer 3: This note can be removed since it's included by the other distinct limit release note (cc: @kaikalur )

Verifier Changes
________________

> * Add support to verifier for running control and test queries concurrently through config `concurrent-control-and-test` and test property.

Reviewer 2: Use double backticks ``

Author: updated

SPI Changes
___________

> * Add `getInfoMap` method in `ConnectorSplit` which returns a `Map<String, String>`. This method should be preferred to the `getInfo` method which returns a raw object.

Reviewer 3: everything in single backticks should be double backtics

Author: corrected.

Hive Changes
____________

> * Fix Cassandra documentation for supported partition types.

Reviewer: Is "Cassandra documentation" for Cassandra or for Hive?

Reviewer 5: I think we can remove documentation related changes from the release notes?

Author: removed

> * Add a limit that prevents caching partitions with column counts greater than a configured threshold. This threshold can be set using the `hive.partition-cache-column-count-limit` configuration parameter.

Reviewer 3: the configuration property should be in double backticks, not single

Author: corrected.

* Add metastore config property ``hive.metastore.thrift.delete-files-on-table-drop`` to delete files on drop table (:pr:`17369`).
* ...

Iceberg Changes
_______________

> * Add $properties system table.

Reviewer: Does $properties need quotes?

Author: updated

> * Add schema evolution in the iceberg doc.

Reviewer 3: We don't need release notes for documentation changes.

Author: done. remove the item "* Add time travel in the iceberg doc." as well

* Add time travel in the iceberg doc.
* Store column comments for iceberg tables.
* Upgrade Iceberg to 0.13.1.

Mongodb Changes
_______________

> * Add CAST(ObjectId() as STRING).

Reviewer: add quotes and :func:?

Author: :func: added

Pinot Changes
_____________

> * Always double-quotes identifiers in generated Pinot queries.

Reviewer 2: This sentence seems to be broken. Also I'm not sure what is the user impact here.

> * Support TLS in Pinot gRPC connection.

Reviewer 2: Add TLS support in Pinot grPC connection

Author: updated

> * Turn on config `pinot.pushdown-topn-broker-queries` by default to push down aggregation group by order by queries to improve query performance.

Reviewer 2: Use double backticks ``

Author: updated.

* Enhance assertQuery to support actual and expected session parameters
* Throw ``NOT_SUPPORTED`` PrestoException if the SQL statement has an offset clause without ``offset_clause_enabled`` being set
* Allow connectors to support column subfield access control checks. Connectors can specify subfield access control for row-type columns through ``checkCanSelectFromColumns()`` in SPI.
* Upgrade postgresql driver to 42.3.3.

Verifier Changes
________________
* Add support to verifier for running control and test queries concurrently through config ``concurrent-control-and-test`` and test property.

SPI Changes
___________
* Add ``getInfoMap`` method in ``ConnectorSplit`` which returns a ``Map<String, String>``. This method should be preferred to the ``getInfo`` method which returns a raw object.

Hive Changes
____________
* Add a limit that prevents caching partitions with column counts greater than a configured threshold. This threshold can be set using the ``hive.partition-cache-column-count-limit`` configuration parameter.
* Add metastore config property ``hive.metastore.thrift.delete-files-on-table-drop`` to delete files on drop table.

> * Support for overwriting existing partitions with a hive config property ``hive.insert-existing-partitions-behavior``. The config supersedes a legacy config ``hive.insert-overwrite-immutable-partitions-enabled``. The new config adds capability of overwriting new partitions for S3

Reviewer 4: Add support for overwriting...

Author: done

* Replace Hive session property ``streaming_aggregation_enabled`` with  ``order_based_execution_enabled``.
* Replace Hive config ``hive.streaming-aggregation-enabled`` with ``hive.order-based-execution-enabled``.
* If Hive files are changed, the modification time can be used to invalidate parquet metadata cache immediately.

Iceberg Changes
_______________
* Add ``$properties`` system table.

> * Store column comments for iceberg tables.

Reviewer 4: Add support for storing column comments for Iceberg tables.

* Upgrade Iceberg to 0.13.1.
* Fix Iceberg $files table in case of column dropping.

Mongodb Changes
_______________
* Add :func:`CAST(ObjectId() as STRING)`.

Pinot Changes
_____________
* Add double-quotes to Pinot generated queries to ensure that reserved keywords are escaped.
* Add TLS support in Pinot gRPC connection.

> * Turn on config ``pinot.pushdown-topn-broker-queries`` by default to push down aggregation group by order by queries to improve query performance.

Reviewer 4: Improve query performance by enabling pushdown of topn broker queries by default.

* Fix an error in :func:`tdigest_agg`.
* Add :func:`to_base32` and :func:`from_base32` varbinary functions.
* Add support for ``TRUNCATE TABLE`` statement.
* Add support to set timeout on the query analyzer runtime to prevent long running query analysis like complex regular expressions. The timeout duration can be set by the configuration property ``planner.query-analyzer-timeout`` or session property ``query_analyzer_timeout``.
* Add :func:`trim_array` function.
* Add a new optimization for showing results for (interactive) distinct limit N as they become available with no buffering.  This can be enabled by setting the session parameter: ``quick_distinct_limit_enabled`` to true.
* Remove state (PageBuilder) in ``ZIP_WITH`` as it was getting clobbered when used in ``REDUCE_AGG`` context.
* Change protoc and grpc version in presto-grpc-api module to adapt to linux aarch64.
* Upgrade zstandard compression to version 1.5.2.2. This can improve about 2% overall CPU with data compressed.
* Add support for local round robin shuffle to reduce the partial distinct limit output size.

> * Add a Presto connector for Clickhouse with support of username/password authentication

Reviewer 2: This should go inside ClickHouse connector section.
https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines#order-of-sections

Author: done

* Enable retrying max requests queued exception
* Enhance assertQuery to support actual and expected session parameters
* Throw ``NOT_SUPPORTED`` PrestoException if the SQL statement has an offset clause without ``offset_clause_enabled`` being set
* Add support for casting timestamp with micro/nanosecond precision

> * Enable warnings for approx_distinct and approx_set with low ``MAX_STANDARD_ERROR``

Reviewer 2: end lines with period

Author: done.

* Upgrade postgresql driver to 42.3.3.

Verifier Changes
________________
* Add support to verifier for running control and test queries concurrently through config ``concurrent-control-and-test`` and test property.

SPI Changes
___________
* Add ``getInfoMap`` method in ``ConnectorSplit`` which returns a ``Map<String, String>``. This method should be preferred to the ``getInfo`` method which returns a raw object.

Hive Changes
____________
* Add a limit that prevents caching partitions with column counts greater than a configured threshold. This threshold can be set using the ``hive.partition-cache-column-count-limit`` configuration parameter.
* Add metastore config property ``hive.metastore.thrift.delete-files-on-table-drop`` to delete files on drop table.
* Support for overwriting existing partitions with a hive config property ``hive.insert-existing-partitions-behavior``. The config supersedes a legacy config ``hive.insert-overwrite-immutable-partitions-enabled``. The new config adds capability of overwriting new partitions for S3
* Replace Hive session property ``streaming_aggregation_enabled`` with  ``order_based_execution_enabled``

> * Replace Hive config ``hive.streaming-aggregation-enabled`` with ``hive.order-based-execution-enabled``

Reviewer 2: end lines with period.

* Fix a bug where ``EXPLAIN (TYPE IO)`` did not include the index source from an index join.
* Fix a memory over-counting issue for using varchar types that could cause queries to fail with exceeded memory limit errors.
* Fix an off-by-one error during buffer size calculation in :func:`tdigest_agg`.
* Fix memory tagging for UnnestOperator and TableWriterMergeOperator.
* Improve performance of :func: `zip_with` when used inside a :func:`reduce_agg`.
* Add :func:`to_base32` and :func:`from_base32` varbinary functions.
* Add support for ``TRUNCATE TABLE`` statement and associated :doc:`/sql/truncate`.
* Add support to set timeout on the query analyzer runtime to prevent long running query analysis like complex regular expressions. The timeout duration can be set by the configuration property ``planner.query-analyzer-timeout`` or session property ``query_analyzer_timeout``.
* Add :func:`trim_array` function that can be called to delete elements from the end of an ordinary array.

> * Add a new optimization for showing results for (interactive) distinct limit N as they become available with no buffering.  This can be enabled by setting the session parameter: ``quick_distinct_limit_enabled`` to true.

Reviewer 4: nit: ....by  setting the session **property** ``quick_distinct_limit_enabled`` to true.

* Add support for casting timestamp with micro/nanosecond precision.
* Add Read support for json & jsonb data types in postgresql.
* Add support for local round robin shuffle to reduce the partial distinct limit output size.
* Add support for column subfield access control checks in connectors. Connectors can specify subfield access control for row-type columns through ``checkCanSelectFromColumns()`` in SPI.
* Add warnings for :func:`approx_distinct` and :func:`approx_set` with low ``MAX_STANDARD_ERROR``.
* Add support for retrying queries that failed with max requests queued exception.
* Update protoc and grpc version in presto-grpc-api module to adapt to linux aarch64.
* Upgrade zstandard compression to version 1.5.2.2. This improves the cpu used for data compressed with zstandard by about 2%.
* Upgrade postgresql driver to 42.3.3.

Clickhouse Changes
____________
* Add a Presto connector for Clickhouse with support of username/password authentication.

Hive Changes
____________

> * Add a limit that prevents caching partitions with column counts greater than a configured threshold. This threshold can be set using the ``hive.partition-cache-column-count-limit`` configuration parameter.

Reviewer 4: ....`hive.partition-cache-column-count-limit`` configuration **property**.

> * Add metastore config property ``hive.metastore.thrift.delete-files-on-table-drop`` to delete files on drop table.

Reviewer 4: nit: `configuration`

> * Add support for overwriting existing partitions with a hive config property ``hive.insert-existing-partitions-behavior``. The config supersedes a legacy config ``hive.insert-overwrite-immutable-partitions-enabled``. The new config adds capability of overwriting new partitions for S3.

Reviewer 4: nit: `configuration`

Iceberg Changes
_______________
* Fix Iceberg ``$files`` table in case of column dropping.
* Add ``$properties`` system table.
* Add support for storing column comments for Iceberg tables.
* Upgrade Iceberg to 0.13.1.

Mongodb Changes
_______________
* Add :func:`CAST(ObjectId() as STRING)`.

Pinot Changes
_____________
* Improve query performance by enabling pushdown of topn broker queries by default.
* Add double-quotes to Pinot generated queries to ensure that reserved keywords are escaped.
* Add TLS support in Pinot gRPC connection.
* Upgrade Pinot connector to support Pinot 0.10.0.

Spark Changes
_____________
* Improve handling of scenario where a query fails due to broadcast table violating node memory limit or broadcast limit. Previously the query failed while workers tried to load broadcast table causing container out of memory.
* Add a new configuration property ``spark.retry-on-out-of-memory-broadcast-join-enabled`` to disable broadcast join on broadcast OOM and retry the query again within the same spark session.  This can be overridden by ``spark_retry_on_out_of_memory_broadcast_join_enabled`` session property.

Verifier Changes
________________

> * Add support to verifier for running control and test queries concurrently through config ``concurrent-control-and-test`` and test property.

Reviewer 4: Add support to verifier for running control and test queries concurrently by setting ``concurrent-control-and-test`` configuration property to true.

=============
Release 0.273
=============

**Highlights**
==============

> 

Author: done

**Details**
===========

General Changes
_______________
* Fix disaggregated coordinator resource group aggregation issue related to ``WAITING_FOR_PREREQUISITE`` queries.
* Fix a bug where ``EXPLAIN (TYPE IO)`` did not include the index source from an index join.
* Fix a memory over-counting issue for using varchar types that could cause queries to fail with exceeded memory limit errors.

> * Fix an error in :func:`tdigest_agg`.

Reviewer 5: Providing more description on what error might help

Author: done

* Add :func:`to_base32` and :func:`from_base32` varbinary functions.
* Add support for ``TRUNCATE TABLE`` statement.

> * Add support to set timeout on the query analyzer runtime to prevent long running query analysis like complex regular expressions. The timeout duration can be set by the configuration property ``planner.query-analyzer-timeout` or session property ``query_analyzer_timeout``.

Reviewer 5: ```suggestion
* Add support to set timeout on the query analyzer runtime to prevent long running query analysis like complex regular expressions. The timeout duration can be set by the configuration property ``planner.query-analyzer-timeout`` or session property ``query_analyzer_timeout``.
```

Author: corrected

* Add :func:`trim_array` function.

> * Add a new optimization for showing results for (interactive) distinct limit N as they become available with no buffering.  This can be enabled using the session param: ``quick_distinct_limit_enabled`` to true.

Reviewer 5: ```suggestion
* Add a new optimization for showing results for (interactive) distinct limit N as they become available with no buffering.  This can be enabled by setting the session parameter ``quick_distinct_limit_enabled`` to true.
```

Author: corrected

* Remove state (PageBuilder) in ``ZIP_WITH`` as it was getting clobbered when used in ``REDUCE_AGG`` context.

> * Change protoc and grpc version in presto-grpc-api module to adapt linux aarch64.

Reviewer 5: ```suggestion
* Change protoc and grpc version in presto-grpc-api module to adapt to linux aarch64.
```

Author: corrected

> * Upgrade zstandard compression to version 1.5.2.2. This can improved about 2% overall CPU with data compressed.

Reviewer 5: ```suggestion
* Upgrade zstandard compression to version 1.5.2.2. This can improve about 2% overall CPU with data compressed.
```

Author: corrected

> * We have an option now to do a local round robin shuffle to reduce the partial distinct limit output size.

Reviewer 5: This needs to be rephrased. 

Add support for local round robin shuffle to reduce the partial distinct limit output size

Author: done

> * Add a Presto connector for Clickhouse

Reviewer 5: Add as new connector changes below

> * Add support for username/password authentication to clickhouse connector

Reviewer 5: Should go into clickhouse connector changes

Author: Move to the item above

* Enable retrying max requests queued exception
* Enhance assertQuery to support actual and expected session parameters

> * Throw ``NOT_SUPPORTED`` PrestoException if the SQL statement has a offset clause without ``offset_clause_enabled`` being set

Reviewer 5: ```suggestion
* Throw ``NOT_SUPPORTED`` PrestoException if the SQL statement has an offset clause without ``offset_clause_enabled`` being set
```

Author: corrected

> * Add support for casting timestamp with micro/nanosecond precision

Reviewer 5: Keep all Additions (Starting with Add) together

> * Issue warnings for approx_distinct and approx_set with low ``MAX_STANDARD_ERROR``

Reviewer 5: ```suggestion
* Enable warnings for approx_distinct and approx_set with low ``MAX_STANDARD_ERROR``
```

Author: Updated

* Add Read support for json & jsonb data types in postgresql.
* Allow connectors to support column subfield access control checks. Connectors can specify subfield access control for row-type columns through ``checkCanSelectFromColumns()`` in SPI.
* Fix memory tagging for UnnestOperator and TableWriterMergeOperator.
* Upgrade postgresql driver to 42.3.3.

Verifier Changes
________________
* Add support to verifier for running control and test queries concurrently through config ``concurrent-control-and-test`` and test property.

SPI Changes
___________
* Add ``getInfoMap`` method in ``ConnectorSplit`` which returns a ``Map<String, String>``. This method should be preferred to the ``getInfo`` method which returns a raw object.

Hive Changes
____________
* Fix Cassandra documentation for supported partition types.
* Add a limit that prevents caching partitions with column counts greater than a configured threshold. This threshold can be set using the ``hive.partition-cache-column-count-limit`` configuration parameter.
* Add metastore config property ``hive.metastore.thrift.delete-files-on-table-drop`` to delete files on drop table.

> * Support for overwriting existing partitions with a hive config property `hive.insert-existing-partitions-behavior`. The config supersedes a legacy config `hive.insert-overwrite-immutable-partitions-enabled`. The new config adds capability of overwriting new partitions for S3

Reviewer 5: ```suggestion
* Support for overwriting existing partitions with a hive config property ``hive.insert-existing-partitions-behavior``. The config supersedes a legacy config ``hive.insert-overwrite-immutable-partitions-enabled``. The new config adds capability of overwriting new partitions for S3
```

Author: corrected

* Replace Hive session property ``streaming_aggregation_enabled`` with  ``order_based_execution_enabled``
* Replace Hive config ``hive.streaming-aggregation-enabled`` with ``hive.order-based-execution-enabled``

> * If Hive files are changed, the modification time can be used to invalidate parquet metadata cache immediately.

Reviewer 5: We should add why to this

Iceberg Changes
_______________
* Add ``$properties`` system table.
* Add time travel in the iceberg doc.
* Store column comments for iceberg tables.
* Upgrade Iceberg to 0.13.1.
* Fix Iceberg $files table in case of column dropping.

Mongodb Changes
_______________
* Add :func:`CAST(ObjectId() as STRING)`.

Pinot Changes
_____________

> * Add double-quotes to Pinto generated queries to ensure that reserved keywords are escaped.

Reviewer 5: ```suggestion
* Add double-quotes to Pinot generated queries to ensure that reserved keywords are escaped.
```

Author: corrected

=============
Release 0.273
=============

**Highlights**
==============

> * Add new Presto connector for Clickhouse

Reviewer 2: End line with period.

**Details**
===========

General Changes
_______________
* Fix disaggregated coordinator resource group aggregation issue related to ``WAITING_FOR_PREREQUISITE`` queries. When a query is moved to WAITING_FOR_PREREQUISITE, resource manager would end up counting them towards running queries. This gives a wrong impression to coordinator about running queries and it would end up running less queries than configured.
* Fix a bug where ``EXPLAIN (TYPE IO)`` did not include the index source from an index join.
* Fix a memory over-counting issue for using varchar types that could cause queries to fail with exceeded memory limit errors.
* Fix an off-by-one error during buffer size calculation in :func:`tdigest_agg`.
* Fix memory tagging for UnnestOperator and TableWriterMergeOperator.
* Improve performance of :func: `zip_with` when used inside a :func:`reduce_agg`.
* Add :func:`to_base32` and :func:`from_base32` varbinary functions.

> * Add support for ``TRUNCATE TABLE`` statement.

Reviewer 4: `Add support for ``TRUNCATE TABLE`` statement and associated :doc:`/sql/truncate`.`

* Add support to set timeout on the query analyzer runtime to prevent long running query analysis like complex regular expressions. The timeout duration can be set by the configuration property ``planner.query-analyzer-timeout`` or session property ``query_analyzer_timeout``.
* Add :func:`trim_array` function that can be called to delete elements from the end of an ordinary array.
* Add a new optimization for showing results for (interactive) distinct limit N as they become available with no buffering.  This can be enabled by setting the session parameter: ``quick_distinct_limit_enabled`` to true.
* Add support for casting timestamp with micro/nanosecond precision.
* Add Read support for json & jsonb data types in postgresql.
* Add support for local round robin shuffle to reduce the partial distinct limit output size.
* Add support for column subfield access control checks in connectors. Connectors can specify subfield access control for row-type columns through ``checkCanSelectFromColumns()`` in SPI.

> * Change protoc and grpc version in presto-grpc-api module to adapt to linux aarch64.

Reviewer 4: ```suggestion
* Update protoc and grpc version in presto-grpc-api module to adapt to linux aarch64.
```

* Upgrade zstandard compression to version 1.5.2.2. This improves the cpu used for data compressed with zstandard by about 2%.
* Upgrade postgresql driver to 42.3.3.
* Enable retrying max requests queued exception.
* Enable warnings for approx_distinct and approx_set with low ``MAX_STANDARD_ERROR``.

> * Enhance assertQuery to support actual and expected session parameters

Reviewer 4: We can remove this. Its not important to list in release notes.

> * Throw ``NOT_SUPPORTED`` PrestoException if the SQL statement has an offset clause without ``offset_clause_enabled`` being set

Reviewer 4: We can remove this. The queries are expected to fail, when do not enable a feature.

Author: done

* Add support for timely Parquet metadata cache invalidation.
* Replace Hive session property ``streaming_aggregation_enabled`` with ``order_based_execution_enabled``.
* Replace Hive config ``hive.streaming-aggregation-enabled`` with ``hive.order-based-execution-enabled``.

Iceberg Changes
_______________
* Fix Iceberg $files table in case of column dropping.
* Add ``$properties`` system table.
* Add support for storing column comments for Iceberg tables.
* Upgrade Iceberg to 0.13.1.

Mongodb Changes
_______________
* Add :func:`CAST(ObjectId() as STRING)`.

Pinot Changes
_____________
* Add double-quotes to Pinot generated queries to ensure that reserved keywords are escaped.
* Add TLS support in Pinot gRPC connection.
* Improve query performance by enabling pushdown of topn broker queries by default.
* Upgrade Pinot connector to support Pinot 0.10.0.

Spark Changes
_____________
* Add a new configuration property ``spark.retry-on-out-of-memory-broadcast-join-enabled`` to disable broadcast join on broadcast OOM and retry the query again within the same spark session.  This can be overridden by ``spark_retry_on_out_of_memory_broadcast_join_enabled`` session property.

> * Improve handling of scenario where a query fails due to broadcast table violating node memory limit or broadcast limit. Previously the query failed while workers tried to load broadcast table causing container out of memory.

Author: done
